Unconstrained type parameters and `??`

BillWagner opened the pull request:
Fixes #737

Add the language to the null-coalescing operator so that `??` can be used with an unconstrained type parameter.

Notes for reviewers:

After looking at the final change, glance at the commit by commit history.

The only change is to remove the phrase "or a reference type" from the first bullet. By doing that, an unconstrained type parameter `T` is allowed. It's not a reference type.

I tested the case Nigel and Kalle described for using `??` with an unconstrained type parameter, and the existing rules produce the behavior they described.

I ran a few other cases to validate, and those worked correctly as well.

The commit-by-commit history shows what I first started to write by adding a bullet for unconstrained type parameters, then spec'ing that behavior. That started me on the path of just removing the restriction. I checked with the compiler team, and they agreed.

BillWagner committed:
Unconstrainted type parameters and `??`

Fixes #737

Add the language to the null-coalescing operator so that `??` can be used with an unconstrained type parameter.

BillWagner committed:
Revert "Unconstrainted type parameters and `??`"

This reverts commit a958a379933840153266ab31ce6e6e0780ace2f5.

BillWagner committed:
Do this the easy way.

I think this can be done the easy way. After reverting the original change, this commit removes the phrase "or a reference type" from the first bullet. By doing that, an unconstrained type parameter `T` is allowed. It's not a reference type.

I tested the case Nigel and Kalle described for using `??` with an unconstrained type parameter, and the existing rules produce the behavior they described.

I ran a few other cases to validate, and those worked correctly as well.

BillWagner commented:
related to #609, but this can be merged separately.

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -4822,7 +4822,7 @@ The null coalescing operator is right-associative, meaning that operations are g
 The type of the expression `a ?? b` depends on which implicit conversions are available on the operands. In order of preference, the type of `a ?? b` is `A₀`, `A`, or `B`, where `A` is the type of `a` (provided that `a` has a type), `B` is the type of `b`(provided that `b` has a type), and `A₀` is the underlying type of `A` if `A` is a nullable value type, or `A` otherwise. Specifically, `a ?? b` ...
 
-- If `A` exists and is not a nullable value type or a reference type, a compile-time error occurs.
+- If `A` exists and is not a nullable value type, a compile-time error occurs.
```
I’ll review this properly later but for now did you mean:

```suggestion
- If `A` exists and is a non-nullable value type a compile-time error occurs.
```

as the current wording would *exclude* reference types? Should other types be invalid, e.g. pointer types?

jskeet replied:
I share the same confusion as Nigel here. And I don't know offhand whether a type parameter that is constrained (e.g. `T where T : struct`) counts as "a non-nullable value type" or just "known to be a non-nullable value type".

BillWagner replied:
Does `T where T : struct` counts as a non-nullable value type.  I think so. Here's a small test:

```csharp
int? number = null;
int? number2 = 4;
IsNullableValueTypeAStruct(number); // Error: CS0453 "The type 'int?' must be a non-nullable value type in order to use it as a parameter 'T' in the generic type or method 'NullCoaelescing.IsNullableValueTypeAStruct<T>(T)'"
NullCoalescingStruct(5, 12);
NullCoalescingStruct(number, number2); // // Error: CS0453

void IsNullableValueTypeAStruct<T>(T a) where T : struct  { }

void NullCoalescingStruct<T>(T a, T b) where T : struct 
{
    T result = a ?? b; // Error: CS0019 "Operator '??' cannot be applied to operands of type 'T' and 'T'"
}
```

BillWagner replied:
I applied Nigel's suggestion and excluded unmanaged types (which includes pointer types). There is some duplication, but it feels more correct.

I unresolved the conversation because of Jon's question.

ericlippert replied:
I concur Bill; if I recall correctly the intention of the C# 2.0 design was that `where T:struct` definitely ensures that `T` is a non-nullable value type.

jskeet reviewed (commented):
I'm hoping this will be simple to get agreed and merged in the meeting.

BillWagner committed:
Update standard/expressions.md

Co-authored-by: Nigel-Ecma <6654683+Nigel-Ecma@users.noreply.github.com>

BillWagner committed:
add exclusion on unmanaged types

Respond to Nigel's question

Nigel-Ecma reviewed (dismissed):
Jon’s question regarding whether “is a X type” includes type parameters constrained to X needs to be addressed. Semantically the answer is yes, but that is not clear in the text.

Further that unconstrained type parameters are allowed is maybe not obvious, or maybe that such an unconstrained type parameters can be instantiated with a non-nullable type at runtime and produce no errors (`a ?? b` in this case is `a` as `a` is not `null`).

I think “maybe not obvious” can probably be covered by note/example with x-refs and not require rewording the list.

BillWagner committed:
respond to feedback

Change "a non-nullable value type" to "known to be a non-nullable value type".

Add an example to illustrate that `a ?? b` is `a` when the type of `a` is a non-nullable value type used as a type argument to an unconstrained type parameter.

BillWagner commented:
> Jon’s question regarding whether “is a X type” includes type parameters constrained to X needs to be addressed. Semantically the answer is yes, but that is not clear in the text.
> 
> Further that unconstrained type parameters are allowed is maybe not obvious, or maybe that such an unconstrained type parameters can be instantiated with a non-nullable type at runtime and produce no errors (`a ?? b` in this case is `a` as `a` is not `null`).
> 
> I think “maybe not obvious” can probably be covered by note/example with x-refs and not require rewording the list.

Updated the text and added an example.

Nigel-Ecma commented on standard/expressions.md:
````diff
@@ -4831,6 +4831,20 @@ The type of the expression `a ?? b` depends on which implicit conversions are
 
 Otherwise, `a` and `b` are incompatible, and a compile-time error occurs.
 
+> *Example*:
+>
+> <!-- Example: {template:"standalone-console-without-using", name:"DeclarationExpressions1"} -->
+> ```csharp
+> T M<T>(T a, T b) => a ?? b;
+>
+> string s = M(null, "text");
+> int i = M(19, 23);
+> ```
+>
+> The parameters to `M` are unconstrained. Therefore, the type arguments can be reference types, or nullable value types, as shown in the first call to `M`. The type arguments can also be non-nullable value types, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
+>
+> *end example*
````
I don’t think this is there yet.

Bullets 2 thru 5 are all of the form “Otherwise, if `A` exists & <cond>… At runtime…”. This is a compile-time condition followed by the runtime actions. If `A` is an unconstrained type parameter then at compile-time it certainly exists (a type-parameter cannot be bound to a non-type), but what it is is not known until runtime. So this gets you to bullet 6, which is also where you end up if A does not exist (i.e. `a` is `null`), and this is incorrect for the dynamic case – it is correct for `null`.

There needs to be a bullet 5A (i.e. an inserted one), which covers runtime resolution and may be defined either from scratch, may include applying 2-5 at runtime, or whatever is needed.

This example can then refer to the rules and say which result is used in each case.

BTW I think line 4832 should be bullet 7.

Sorry, no time to attempt a starting point for this right now.

BillWagner replied:
I did add line 4832 to the bullet list.

The other concerns are ones I'd like to address in a new PR. The condition Nigel describes was already in the standard. A different task for this part makes sense to me.

jskeet commented on standard/expressions.md:
````diff
@@ -4831,6 +4831,20 @@ The type of the expression `a ?? b` depends on which implicit conversions are
+> int i = M(19, 23);
+> ```
+>
+> The parameters to `M` are unconstrained. Therefore, the type arguments can be reference types, or nullable value types, as shown in the first call to `M`. The type arguments can also be non-nullable value types, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
````
```suggestion
> The parameters to `M` are unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
```

(Just because there's only a single type parameter, and therefore a single type argument.)

jskeet replied:
For "the expression `a ?? b` is `a`" should we have "is evaluated equivalently to" or something similar?

Nigel-Ecma replied:
@jskeet – No, if `a` is a non-nullable value type then `a` is non-null (think of the test as `(object)a == null` which for an IL target is a `box` and test) and so `a` is the result – it is not different handling *per se*. What is needed is to revise the bulleted list to cover this, see my earlier comment.

jskeet replied:
My point is that the "is" here is a bit ambiguous. Perhaps "the value of the expression `a ?? b` is the value of `a`" or something like that? Just "is" seems wrong to me.

Nigel-Ecma replied:
@jskeet – I see, your "is evaluated equivalently to" threw me and “is the value of” is a lot better. But that leaves us with the problem that this example is not supported by the text, we could say its aspirational! ;-)

BillWagner committed:
Update standard/expressions.md

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner committed:
extend the bulleted list.

BillWagner committed:
Merge branch 'unconstrained-null-coalescing' of https://github.com/BillWagner/csharpstandard into unconstrained-null-coalescing

BillWagner committed:
fix test harness issue

jskeet commented on standard/expressions.md:
````diff
@@ -4822,14 +4822,27 @@ The null coalescing operator is right-associative, meaning that operations are g
+> int i = M(19, 23);
+> ```
+>
+> The parameters to `M` are unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
````
```suggestion
> The type parameter `T` for method `M` is unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
```

jnm2 commented on standard/expressions.md:
```diff
@@ -4822,14 +4822,27 @@ The null coalescing operator is right-associative, meaning that operations are g
 The type of the expression `a ?? b` depends on which implicit conversions are available on the operands. In order of preference, the type of `a ?? b` is `A₀`, `A`, or `B`, where `A` is the type of `a` (provided that `a` has a type), `B` is the type of `b`(provided that `b` has a type), and `A₀` is the underlying type of `A` if `A` is a nullable value type, or `A` otherwise. Specifically, `a ?? b` ...
 
-- If `A` exists and is not a nullable value type or a reference type, a compile-time error occurs.
+- If `A` exists and is an unmanaged type (§8.8) or known to be a non-nullable value type, a compile-time error occurs.
```
Does "non-nullable value type" not subsume "unmanaged type"?

jnm2 replied:
Ah, I missed the pointers conversation.

jskeet committed:
Update standard/expressions.md

jskeet commented on standard/expressions.md:
````diff
@@ -4822,14 +4822,27 @@ The null coalescing operator is right-associative, meaning that operations are g
+> int i = M(19, 23);
+> ```
+>
+> The type parameter `T` for method `M` is unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
````
```suggestion
> The type parameter `T` for method `M` is unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the value of the expression `a ?? b` is `a`.
```

jskeet committed:
Update standard/expressions.md
